examples/training/object_detection_3d/waymo/train_pillars.py

```diff
@@ -1,129 +1,320 @@
     )
-    boxes = boxes.to_tensor(
-        default_value=-1.0, shape=[global_batch, 1, 1000, 11]
+
+    """
+    Next, we'll need a decoder component to decode predictions into 3D boxes. To
+    do this, we'll need to specify how many heading bins we're using for each
+    class, the anchor size for each class, and a pooling size for each class.
+    """
+
+    # 12 heading bins for cars, 4 for pedestrians.
+    num_heading_bins = [12, 4]
+
+    decoder = MultiClassHeatmapDecoder(
+        num_classes=2,
+        num_head_bin=num_heading_bins,
+        anchor_size=[car_anchor_size, pedestrian_anchor_size],
+        max_pool_size=[7, 3],
+        max_num_box=[800, 400],
+        heatmap_threshold=[0.1, 0.1],
+        voxel_size=voxel_size,
+        spatial_size=spatial_size,
     )
-    res["point_clouds"] = point_clouds
-    res["bounding_boxes"] = boxes
-    return res
 
+    """
+    Finally, we'll create a detection head and then instantiate our full model.
+    Now we can compile the model and start training!
+    """
+    multiclass_head = MultiClassDetectionHead(
+        num_classes=2,
+        num_head_bin=num_heading_bins,
+    )
+
+    model = MultiHeadCenterPillar(
+        backbone=CenterPillarBackbone.from_preset(
+            "center_pillar_waymo_open_dataset"
+        ),
+        voxel_net=voxelization_layer,
+        multiclass_head=multiclass_head,
+        prediction_decoder=decoder,
+    )
+
+    return model
+
+
+"""
+Before we start training our model, we'll need to turn our labels into a format
+that our model can learn and later predict.
+
+We do this using a label encoder (much like we do in 2D object detection).
+"""
+
+label_encoder = CenterNetLabelEncoder(
+    voxel_size=voxel_size,
+    max_radius=[8.0, 8.0, 0],
+    spatial_size=spatial_size,
+    num_classes=2,
+    # The maximum number of target boxes that we should produce per class
+    # (in this case 1024 for cars and 512 for pedestrians)
+    top_k_heatmap=[1024, 512],
+)
+
+dataset = dataset.map(label_encoder, num_parallel_calls=tf.data.AUTOTUNE)
+
+
+# Up to this point, our data has been in one dictionary per-batch, but
+# now we split it up into a standard x, y tuple for training
+def separate_points_and_boxes(y):
+    x = y["point_clouds"]
+    del y["point_clouds"]
 
-# Load the training dataset
-filenames = tf.data.Dataset.list_files(
-    os.path.join(TRAINING_RECORD_PATH, "*.tfrecord")
+    return x, y
+
+
+dataset = dataset.map(
+    separate_points_and_boxes, num_parallel_calls=tf.data.AUTOTUNE
 )
-train_ds = tf.data.TFRecordDataset(filenames)
-train_ds = train_ds.map(build_tensors, num_parallel_calls=tf.data.AUTOTUNE)
-train_ds = train_ds.apply(
-    tf.data.experimental.dense_to_ragged_batch(
-        global_batch, drop_remainder=True
+
+"""
+Now we can build and compile our model!
+We use a one device strategy in this tutorial, but any strategy will work.
+"""
+
+strategy = tf.distribute.OneDeviceStrategy("/gpu:0")
```

Comment:
Why is a strategy needed for one device? Don't we do this all the time in our examples without this complication?